File: C/6.ZigzagConversion.c
//Original code: for the second layer of for loop, the step of j should be T instead of 1, which can significantly reduce the time complexity from O(n^2) to O(n).
char* convert(char* s, int numRows) {
    int n = strlen(s), k = 0, T = 2 * numRows - 2;
    if (numRows == 1 || numRows >= n) return s;
    char* result = (char*)malloc((n + 1) * sizeof(char));
    for (int i = 0; i < numRows; i++) {
        for (int j = 0; j < n; j++) {
            if ((j - i) % T == 0 || (j + i) % T == 0) {
                result[k] = s[j];
                k++;
            }
        }
    }
    result[n] = '\0';
    return result;
}
//My new code:
char* convert(char* s, int numRows) {
    int n=strlen(s), k=0, T=2*numRows-2;
    if(numRows==1 || numRows>=n) return s;
    char* result=(char*)malloc((n+1)*sizeof(char));
    for(int row=0;row<numRows;row++){
        for(int j=row;j<n;j+=T){
            result[k]=s[j];
            k++;
            if(row>0 && row<numRows-1 && j+T-2*row < n){
                result[k]=s[j+T-2*row];
                k++;
            }
        }
    }
    result[n]='\0';
    return result;
}

File: Python/6.ZigzagConversion.py
#awaiting for improvement...
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        n,k,T=len(s),0,2*numRows-2
        if numRows==1 or numRows>=n: return s
        result=''
        for row in range(0,numRows):
            for j in range(row,n,T):
                result+=s[j]
                if row>0 and row<numRows-1 and j+T-2*row<n:
                    result+=s[j+T-2*row]
        return result